clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -7,24 +7,58 @@ import io.treeverse.clients.LakeFSContext.{
   LAKEFS_CONF_API_URL_KEY
 }
 import org.apache.hadoop.conf.Configuration
+import org.apache.spark.broadcast.Broadcast
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.types.{StringType, StructField, StructType}
 import org.apache.spark.sql.{SparkSession, _}
-import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration
-import software.amazon.awssdk.core.retry.RetryPolicy
-import software.amazon.awssdk.core.retry.backoff.BackoffStrategy
-import software.amazon.awssdk.regions.Region
-import software.amazon.awssdk.services.s3.S3Client
-import software.amazon.awssdk.services.s3.model.{Delete, DeleteObjectsRequest, ObjectIdentifier}
+
+import com.amazonaws.services.s3.AmazonS3
+import com.amazonaws.services.s3.model
 
 import java.net.URI
 import collection.JavaConverters._
 
+/** Interface to build an S3 client.  The object
+ *  io.treeverse.clients.conditional.S3ClientBuilder -- conditionally
+ *  defined in a separate file according to the supported Hadoop version --
+ *  implements this trait.  (Scala requires companion objects to be defined
+ *  in the same file, so it cannot be a companion.)
+ */
+trait S3ClientBuilder extends Serializable {
+
+  /** Return a configured Amazon S3 client similar to the one S3A would use.
+   *  On Hadoop versions >=3, S3A can assume a role, and the returned S3
+   *  client will similarly assume that role.
+   *
+   *  @param hc (partial) Hadoop configuration of fs.s3a.
+   *  @param bucket that this client will access.
+   *  @param region to find this bucket.
+   *  @param numRetries number of times to retry on AWS.
+   */
+  def build(hc: Configuration, bucket: String, region: String, numRetries: Int): AmazonS3
+}
+
 object GarbageCollector {
+  type ConfMap = List[(String, String)]
 
   case class APIConfigurations(apiURL: String, accessKey: String, secretKey: String)
 
+  /** @return a serializable summary of values in hc starting with prefix.
+   */
+  def getHadoopConfigurationValues(hc: Configuration, prefix: String): ConfMap =
+    hc.iterator.asScala
+      .filter(_.getKey.startsWith(prefix))
+      .map(e => (e.getKey, e.getValue))
```

Comment:
nit; 
```suggestion
      .map(c => (c.getKey, c.getValue))
```

Reply:
Not done.  This is an iteration over `Map.entry`'s, `e` is traditional Java there.  Changing to use `entry` instead to make that clear.

Reply:
makes sense, thanks